File: agent/packages/opencode/src/video/runstate.ts
import * as fs from "fs/promises"
import * as path from "path"
import type { VideoPayload } from "./payload"

export interface VideoRunState {
  status: string
  promptPath?: string
  runDir?: string
  videoURL?: string
  error?: string
  updatedAt: string
}

export function defaultRunDir(promptPath: string, now = new Date()): string {
  const stamp = [
    now.getFullYear().toString().padStart(4, "0"),
    (now.getMonth() + 1).toString().padStart(2, "0"),
    now.getDate().toString().padStart(2, "0"),
    "-",
    now.getHours().toString().padStart(2, "0"),
    now.getMinutes().toString().padStart(2, "0"),
    now.getSeconds().toString().padStart(2, "0"),
  ].join("")
  return path.join(path.dirname(path.resolve(promptPath)), "runs", stamp)
}

async function exists(filePath: string): Promise<boolean> {
  try {
    await fs.stat(filePath)
    return true
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") return false
    throw err
  }
}

export async function createRunDir(promptPath: string, requested?: string): Promise<string> {
  let runDir = requested ? path.resolve(requested) : defaultRunDir(promptPath)
  if (!requested) {
    const base = runDir
    for (let suffix = 1; await exists(runDir); suffix++) {
      runDir = `${base}-${suffix.toString().padStart(2, "0")}`
    }
  } else if (await exists(runDir)) {
    const entries = await fs.readdir(runDir)
    if (entries.length > 0) throw new Error(`run directory already exists and is not empty: ${runDir}`)
  }
  await fs.mkdir(runDir, { recursive: true })
  return runDir
}

export async function writeJSON(filePath: string, value: unknown): Promise<void> {
  await fs.mkdir(path.dirname(filePath), { recursive: true })
  await fs.writeFile(filePath, `${JSON.stringify(value, null, 2)}\n`)
}

export async function writeDryRun(runDir: string, promptPath: string, request: VideoPayload): Promise<VideoRunState> {
  await writeJSON(path.join(runDir, "request.json"), request)
  const state: VideoRunState = {
    status: "dry_run",
    promptPath: path.resolve(promptPath),
    runDir,
    updatedAt: new Date().toISOString(),
  }
  await writeJSON(path.join(runDir, "state.json"), state)
  return state
}

export async function readState(runDir: string): Promise<VideoRunState> {
  const raw = await fs.readFile(path.join(runDir, "state.json"), "utf8")
  return JSON.parse(raw) as VideoRunState
}

File: video-agent-test/internal/runstate/runstate.go
package runstate

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"time"
)

type State struct {
	Status      string `json:"status"`
	PromptPath  string `json:"prompt_path,omitempty"`
	RunDir      string `json:"run_dir,omitempty"`
	VideoURL    string `json:"video_url,omitempty"`
	Error       string `json:"error,omitempty"`
	SubmittedAt string `json:"submitted_at,omitempty"`
	UpdatedAt   string `json:"updated_at"`
}

func DefaultRunDir(promptPath string, now time.Time) string {
	return filepath.Join(filepath.Dir(promptPath), "runs", now.Format("20060102-150405"))
}

func CreateRunDir(promptPath string, requested string, now time.Time) (string, error) {
	runDir := requested
	if runDir == "" {
		runDir = DefaultRunDir(promptPath, now)
		base := runDir
		for suffix := 1; exists(runDir); suffix++ {
			runDir = fmt.Sprintf("%s-%02d", base, suffix)
		}
	} else if exists(runDir) {
		entries, err := os.ReadDir(runDir)
		if err != nil {
			return "", err
		}
		if len(entries) > 0 {
			return "", fmt.Errorf("run directory already exists and is not empty: %s", runDir)
		}
	}
	if err := os.MkdirAll(runDir, 0o755); err != nil {
		return "", err
	}
	return filepath.Abs(runDir)
}

func LatestRunDir(promptPath string) (string, bool, error) {
	runsDir := filepath.Join(filepath.Dir(promptPath), "runs")
	entries, err := os.ReadDir(runsDir)
	if err != nil {
		if os.IsNotExist(err) {
			return "", false, nil
		}
		return "", false, err
	}

	var names []string
	for _, entry := range entries {
		if entry.IsDir() {
			names = append(names, entry.Name())
		}
	}
	if len(names) == 0 {
		return "", false, nil
	}
	sort.Strings(names)
	return filepath.Join(runsDir, names[len(names)-1]), true, nil
}

func ReadState(runDir string) (*State, error) {
	raw, err := os.ReadFile(filepath.Join(runDir, "state.json"))
	if err != nil {
		return nil, err
	}
	var state State
	if err := json.Unmarshal(raw, &state); err != nil {
		return nil, err
	}
	return &state, nil
}

func WriteState(runDir string, state State) error {
	state.RunDir = runDir
	state.UpdatedAt = time.Now().UTC().Format(time.RFC3339)
	return WriteJSON(filepath.Join(runDir, "state.json"), state)
}

func WriteJSON(path string, value any) error {
	raw, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		return err
	}
	raw = append(raw, '\n')
	return WriteFileAtomic(path, raw, 0o644)
}

func WriteString(path string, value string) error {
	return WriteFileAtomic(path, []byte(value), 0o644)
}

func AppendJSONL(path string, value any) error {
	raw, err := json.Marshal(value)
	if err != nil {
		return err
	}
	raw = append(raw, '\n')
	file, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	defer file.Close()
	_, err = file.Write(raw)
	return err
}

func WriteFileAtomic(path string, data []byte, perm os.FileMode) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	tmp := fmt.Sprintf("%s.tmp.%d", path, time.Now().UnixNano())
	if err := os.WriteFile(tmp, data, perm); err != nil {
		return err
	}
	return os.Rename(tmp, path)
}

func IsTerminal(status string) bool {
	switch status {
	case "succeeded", "failed", "timeout", "dry_run":
		return true
	default:
		return false
	}
}

func exists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}
